src/geo/projection/globe.js

```diff
@@ -0,0 +1,506 @@
+        this.poleSegments = SegmentVector.simpleSegment(0, 0, poleVertices, polePrimitives);
+
+        const atmosphereVertices = new GlobeVertexArray();
+        atmosphereVertices.emplaceBack(-1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 0.0);
+        atmosphereVertices.emplaceBack(1.0, 1.0, 1.0, 0.0, 0.0, 1.0, 0.0);
+        atmosphereVertices.emplaceBack(1.0, -1.0, 1.0, 0.0, 0.0, 1.0, 1.0);
+        atmosphereVertices.emplaceBack(-1.0, -1.0, 1.0, 0.0, 0.0, 0.0, 1.0);
+
+        const atmosphereTriangles = new TriangleIndexArray();
+        atmosphereTriangles.emplaceBack(0, 1, 2);
+        atmosphereTriangles.emplaceBack(2, 3, 0);
+
+        this.atmosphereVertexBuffer = context.createVertexBuffer(atmosphereVertices, atmosphereLayout.members);
+        this.atmosphereIndexBuffer = context.createIndexBuffer(atmosphereTriangles);
+        this.atmosphereSegments = SegmentVector.simpleSegment(0, 0, 4, 2);
+    }
+
+    destroy() {
+        this.poleIndexBuffer.destroy();
+        this.gridIndexBuffer.destroy();
+        this.poleSegments.destroy();
+        this.gridSegments.destroy();
+        this.atmosphereVertexBuffer.destroy();
+        this.atmosphereIndexBuffer.destroy();
+        this.atmosphereSegments.destroy();
+
+        if (this.wireframeIndexBuffer) {
+            this.wireframeIndexBuffer.destroy();
+            this.wireframeSegments.destroy();
+        }
+    }
+
+    static createPoleTriangleVertices(tiles: number, ws: number, isTopCap: boolean): GlobeVertexArray {
+        const lerp = (a, b, t) => a * (1 - t) + b * t;
+        const arr = new GlobeVertexArray();
+        const radius = ws / Math.PI / 2.0;
+
+        // Place the tip
+        arr.emplaceBack(0, -radius, 0, 0, 0, 0.5, isTopCap ? 0.0 : 1.0);
+
+        const startAngle = 0;
+        const endAngle = 360.0 / tiles;
+        const cosLat = Math.cos(degToRad(85.0));
+        const sinLat = Math.sin(degToRad(85.0));
+
+        for (let i = 0; i <= GLOBE_VERTEX_GRID_SIZE; i++) {
+            const uvX = i / GLOBE_VERTEX_GRID_SIZE;
+            const angle = lerp(startAngle, endAngle, uvX);
+            const p = csLatLngToECEF(cosLat, sinLat, angle, radius);
+
+            arr.emplaceBack(p[0], p[1], p[2], 0, 0, uvX, isTopCap ? 0.0 : 1.0);
+        }
+
+        return arr;
+    }
+
+    _createPoleTriangleIndices(): TriangleIndexArray {
+        const arr = new TriangleIndexArray();
+        for (let i = 0; i <= GLOBE_VERTEX_GRID_SIZE; i++) {
+            arr.emplaceBack(0, i + 1, i + 2);
+        }
+        return arr;
+    }
+
+    static createGridVertices(sx: number, sy: number, sz: number): GlobeVertexArray {
+        const lerp = (a, b, t) => a * (1 - t) + b * t;
+        const tiles = Math.pow(2, sz);
+        const gridTileId = new CanonicalTileID(sz, sx, sy);
+        const [latLngTL, latLngBR] = globeTileLatLngCorners(gridTileId);
+        const boundsArray = new GlobeVertexArray();
+
+        const bounds = globeTileBounds(new CanonicalTileID(sz, tiles / 2, sy));
```

Comment:
Could we just use `CanonicalTileID(sz, sx, sy)` now that the mesh sharing is not implemented yet? This would simplify the `globeMatrixForTile` function too. Am I missing something?

Reply:
Yes, fixed while addressing https://github.com/mapbox/mapbox-gl-js/pull/11329#discussion_r768675240.